crates/red_knot/src/ast_ids.rs

```diff
@@ -274,6 +274,14 @@ pub struct TypedNodeKey<N: AstNode> {
 }
 
 impl<N: AstNode> TypedNodeKey<N> {
+    pub fn from_node(node: &N) -> Self {
+        let inner = NodeKey {
+            kind: node.as_any_node_ref().kind(),
+            range: node.range(),
+        };
+        Self::new(inner).unwrap()
+    }
+
```

Comment:
Nit: I would construct the key here directly without calling `new` because it should be guaranteed that `N::can_cast(node.as_any_node_ref().kind())` always returns `true`, so the check seems unnecessary. It is probably something that the compiler can optimize away, but that's not guaranteed.
```suggestion
    pub fn from_node(node: &N) -> Self {
        let inner = NodeKey {
            kind: node.as_any_node_ref().kind(),
            range: node.range(),
        };
        Self { inner, _marker: PhantomData }
    }

```